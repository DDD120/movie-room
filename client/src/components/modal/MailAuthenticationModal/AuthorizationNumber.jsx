import styled from "@emotion/styled";
import { colors, fontSize } from "styles/common";

const Base = styled.div`
  position: relative;
  margin: 20px 0;
`;

const Input = styled.input`
  width: 80%;
  padding: 10px 10px;
  font-weight: 700;
  border: none;
  border-bottom: 1px solid #000;
  color: ${colors.black};
  background-color: ${colors.beige};
  outline: none;
  font-size: ${fontSize.lg};
`;

const TimeLimit = styled.span`
  position: absolute;
  top: 50%;
  right: 8px;
  transform: translateY(-50%);
  color: red;
`;

const AuthorizationNumber = ({ timeLimit, inputRef }) => {
  const formatTime = (time) => {
    const min = String(parseInt(time / 60)).padStart(2, "0");
    const sec = String(time % 60).padStart(2, "0");

    return `${min}:${sec}`;
  };
  return (
    <Base>
      <Input
        ref={inputRef}
        title="인증코드"
        type="text"
        min={1}
        max={6}
        placeholder="인증코드"
      />
      <TimeLimit>{formatTime(timeLimit)}</TimeLimit>
    </Base>
  );
};

export default AuthorizationNumber;
