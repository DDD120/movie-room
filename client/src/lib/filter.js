import dayjs from "dayjs";

export const arrayDeduplication = (list) => {
  return list.filter(
    (item, index, array) =>
      index === array.findIndex((data) => data.id === item.id)
  );
};

export const getYear = (date) => {
  return dayjs(date).year();
};
