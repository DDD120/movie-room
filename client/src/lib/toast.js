import { toast } from "react-toastify";

export const showToast = (message) => {
  toast(message, {
    autoClose: 3000,
    hideProgressBar: true,
  });
};
