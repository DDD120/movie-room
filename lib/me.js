export const me = (user) => {
  return {
    id: user._id,
    profile: {
      nickname: user.profile.nickname,
      thumbnail: user.profile.thumbnail,
    },
    email: user.email,
    reviews: user.reviews,
  };
};
