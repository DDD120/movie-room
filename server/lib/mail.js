import nodemailer from "nodemailer";
import dotenv from "dotenv";

dotenv.config();

export const mailSender = async (email, certificationNumber) => {
  let transporter = nodemailer.createTransport({
    service: "gmail",
    host: "smtp.gmail.com",
    port: 587,
    secure: false,
    auth: {
      user: process.env.NODEMAILER_USER,
      pass: process.env.NODEMAILER_PASS,
    },
  });

  try {
    let info = await transporter.sendMail({
      from: `"MOVIE ROOM" <${process.env.NODEMAILER_USER}>`,
      to: email,
      subject: "[MOVIE ROOM] 회원가입 인증번호 발송",
      html: `
        <body>
          <h1>[MOVIE ROOM]</h1>
          </br>
          <p>MOVIE ROOM에서 인증번호를 발송해 드립니다.</p>
          <div style="padding: 12px 0;">
            <span style="font-size: 24px; font-weight; bold;">${certificationNumber}</span>
          </div>
        </body>
      `,
    });

    return info.accepted;
  } catch (e) {
    throw new Error("메일 전송에 실패하였습니다.", console.error(e));
  }
};
