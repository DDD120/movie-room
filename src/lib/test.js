import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import userReducer from "store/user";
import { BrowserRouter } from "react-router-dom";
import { serverApi } from "apis/server-api";
import { moviedbApi } from "apis/movie-db-api";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: {
        user: userReducer,
        [serverApi.reducerPath]: serverApi.reducer,
        [moviedbApi.reducerPath]: moviedbApi.reducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
