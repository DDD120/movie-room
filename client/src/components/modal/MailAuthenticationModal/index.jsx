import { GoMailRead } from "react-icons/go";
import styled from "@emotion/styled";
import { colors, fontSize } from "styles/common";
import { useRef, useEffect } from "react";
import Button from "components/common/Button";
import Modal from "../Modal";
import { useEmailMutation, useSignupMutation } from "apis/server-api";
import AuthorizationNumber from "./AuthorizationNumber";
import useTimer from "hooks/useTimer";
import { useDispatch } from "react-redux";
import { login } from "store/user";
import { showToast } from "lib/toast";
import useAuthRedirect from "hooks/useAuthRedirect";

const Base = styled.main`
  padding: 40px 12px;
`;

const Head = styled.h1`
  font-size: ${fontSize.xl};
  margin-bottom: 10px;
`;

const Explanation = styled.p`
  width: 80%;
  margin: 0 auto;
  word-break: keep-all;
`;

const MailIcon = styled.div`
  color: ${colors.black};
  font-size: ${fontSize.xl};
`;

const MailAuthenticationModal = ({ email, password, onClose }) => {
  const { timeLimit, isRunning, reset } = useTimer();
  const dispatch = useDispatch();
  const inputRef = useRef();
  useAuthRedirect();

  const [
    emailTrigger,
    {
      data: emailRes = {},
      isSuccess: isEmailSuccess,
      isError: isEmailError,
      error: emailError,
    },
  ] = useEmailMutation();
  const [
    signupTrigger,
    {
      data: signupRes = {},
      isSuccess: isSignupSuccess,
      isError: isSignupError,
      error: signupError,
    },
  ] = useSignupMutation();

  const handleCheckClick = () => {
    signupTrigger({
      email,
      password,
      inputNumber: inputRef.current.value,
    });
  };

  const handleResendMailClick = () => {
    emailTrigger({ email });
    reset();
  };

  useEffect(() => {
    if (isEmailSuccess) showToast(emailRes?.message);
    if (isEmailError) showToast(emailError.data?.message);
    if (isSignupError) showToast(signupError.data?.message);
  }, [
    emailRes,
    isEmailSuccess,
    isEmailError,
    emailError,
    isSignupError,
    signupError,
  ]);

  useEffect(() => {
    if (isSignupSuccess) {
      showToast(signupRes?.message);
      onClose();
      dispatch(login({ user: signupRes.user }));
    }
  }, [isSignupSuccess, signupRes, onClose, dispatch]);

  return (
    <Modal onClose={onClose}>
      <Base>
        <MailIcon>
          <GoMailRead />
        </MailIcon>
        <Head>이메일 인증</Head>
        <Explanation>
          이메일 인증을 위한 코드가 <b>{email}</b>으로 전송되었습니다.
        </Explanation>
        <AuthorizationNumber timeLimit={timeLimit} inputRef={inputRef} />
        {isRunning ? (
          <Button onClick={handleCheckClick}>확인</Button>
        ) : (
          <Button onClick={handleResendMailClick}>인증 메일 재전송</Button>
        )}
      </Base>
    </Modal>
  );
};

export default MailAuthenticationModal;
