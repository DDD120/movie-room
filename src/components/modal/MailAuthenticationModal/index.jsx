import { GoMailRead } from "react-icons/go";
import styled from "@emotion/styled";
import { colors } from "styles/common";
import { useRef, useEffect } from "react";
import Button from "components/common/Button";
import Modal from "../Modal";
import { useEmailMutation, useSignupMutation } from "apis/server-api";
import AuthorizationNumber from "./AuthorizationNumber";
import useTimer from "hooks/useTimer";
import { useDispatch } from "react-redux";
import { login } from "store/user";
import { useNavigate } from "react-router-dom";
import { showToast } from "lib/toast";

const Base = styled.main`
  padding: 40px 12px;
`;

const Head = styled.h1`
  font-size: 2rem;
  margin-bottom: 10px;
`;

const Explanation = styled.p`
  width: 80%;
  margin: 0 auto;
  word-break: keep-all;
`;

const MailIcon = styled.div`
  color: ${colors.black};
  font-size: 3rem;
`;

const MailAuthenticationModal = ({ email, password, onClose }) => {
  const { timeLimit, isRunning, reset } = useTimer();
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const inputRef = useRef();
  const [
    emailTrigger,
    {
      data: emailRes = {},
      isSuccess: isEmailSuccess,
      isError: isEmailError,
      error: emailError,
    },
  ] = useEmailMutation();
  const [
    signupTrigger,
    {
      data: signupRes = {},
      isSuccess: isSignupSuccess,
      isError: isSignupError,
      error: signupError,
    },
  ] = useSignupMutation();

  const handleCheckClick = () => {
    signupTrigger({
      email,
      password,
      inputNumber: inputRef.current.value,
    });
  };

  const handleResendMailClick = () => {
    emailTrigger({ email });
    reset();
  };

  useEffect(() => {
    if (isEmailSuccess) {
      showToast(emailRes.message);
    }
  }, [emailRes, isEmailSuccess]);

  useEffect(() => {
    if (isEmailError) {
      showToast(emailError.data.message);
    }
  }, [isEmailError, emailError]);

  useEffect(() => {
    if (isSignupSuccess) {
      showToast(signupRes.message);
      onClose();
      navigate("/");
      dispatch(login({ user: signupRes.user }));
    }
  }, [isSignupSuccess, signupRes, onClose, navigate, dispatch]);

  useEffect(() => {
    if (isSignupError) {
      showToast(signupError.data.message);
    }
  }, [isSignupError, signupError]);

  return (
    <Modal onClose={onClose} backdropTouchClose={false}>
      <Base>
        <MailIcon>
          <GoMailRead />
        </MailIcon>
        <Head>이메일 인증</Head>
        <Explanation>
          이메일 인증을 위한 코드가 <b>{email}</b>으로 전송되었습니다.
        </Explanation>
        <AuthorizationNumber timeLimit={timeLimit} inputRef={inputRef} />
        {isRunning ? (
          <Button clickEvent={handleCheckClick}>확인</Button>
        ) : (
          <Button clickEvent={handleResendMailClick}>인증 메일 재전송</Button>
        )}
      </Base>
    </Modal>
  );
};

export default MailAuthenticationModal;
