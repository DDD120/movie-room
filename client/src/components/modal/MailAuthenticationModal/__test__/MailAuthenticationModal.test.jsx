import { renderWithProviders } from "lib/test";
import MailAuthenticationModal from "..";
import ReactDOM from "react-dom";
import { act, screen } from "@testing-library/react";

jest.useFakeTimers();

describe("MailAuthentication Modal", () => {
  const setup = () => {
    return renderWithProviders(
      <MailAuthenticationModal
        email="email1234@mock.com"
        password="password1234"
        onClose={() => jest.fn()}
      />
    );
  };

  const oldCreatePortal = ReactDOM.createPortal;

  beforeEach(() => {
    ReactDOM.createPortal = (node) => node;
  });

  afterAll(() => {
    ReactDOM.createPortal = oldCreatePortal;
  });

  it("renders correctly", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("should change after 180s the text of the button", async () => {
    setup();

    const checkButton = screen.getByText("확인");
    expect(checkButton).toBeInTheDocument();

    act(() => jest.advanceTimersByTime(179000));

    act(() => jest.advanceTimersByTime(1000));

    const resendButton = await screen.findByText("인증 메일 재전송");
    expect(resendButton).toBeInTheDocument();
  });
});
