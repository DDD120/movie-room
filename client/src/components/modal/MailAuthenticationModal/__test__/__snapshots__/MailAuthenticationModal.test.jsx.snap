// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailAuthentication Modal renders correctly 1`] = `
<div>
  <div
    class="css-glk9o1"
  >
    <div
      class="css-bt0smj"
    >
      <main
        class="css-ozqi8c"
      >
        <div
          class="css-810ok4"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 14 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 5H4V4h2v1zm3 1H4v1h5V6zm5-.48V14c0 .55-.45 1-1 1H1c-.55 0-1-.45-1-1V5.52c0-.33.16-.63.42-.81L2 3.58V3c0-.55.45-1 1-1h1.2L7 0l2.8 2H11c.55 0 1 .45 1 1v.58l1.58 1.13c.27.19.42.48.42.81zM3 7.5L7 10l4-2.5V3H3v4.5zm-2 6l4.5-3-4.5-3v6zm11 .5l-5-3-5 3h10zm1-6.5l-4.5 3 4.5 3v-6z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <h1
          class="css-xvhs7d"
        >
          이메일 인증
        </h1>
        <p
          class="css-1qxvz04"
        >
          이메일 인증을 위한 코드가 
          <b>
            email1234@mock.com
          </b>
          으로 전송되었습니다.
        </p>
        <div
          class="css-1w6xk43"
        >
          <input
            class="css-1orb6qj"
            max="6"
            min="1"
            placeholder="인증코드"
            title="인증코드"
            type="text"
          />
          <span
            class="css-jzs5wp"
          >
            03:00
          </span>
        </div>
        <button
          class="css-19gbx2v"
        >
          확인
        </button>
      </main>
      <button
        class="css-1sa6jya"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
